graph/graph.go

```diff
@@ -0,0 +1,251 @@
+package graph
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Graph is a map-basd adjacency list for storing a lineage graph
+type Graph struct {
+	Graph map[IsID][]IsID
+	Nodes map[IsID]Node
+}
+
+// IDfactory is an interface which creates new node IDs
+type IDfactory interface {
+	New() IsID
+}
+
+// IsID represents any type which can function as a node ID
+type IsID interface {
+	isID()
+}
+
+// autoIncFactory is a factory for an auto-incrementing intID
```

Comment:
Does this need to be thread-safe?

Reply:
No parallelization optimizations are planned.